### PR TITLE
Mark daemons unavailable if the response the client gets back is nonsense

### PR DESCRIPTION
**Pull request by big-guy**
This scenario blocks all new builds:
1. Daemon starts with port 12345
2. Daemon is killed without shutting down cleanly (kill -9 $PID)
3. A non-daemon process starts and listens on port 12345

The Gradle client will see an entry for an idle daemon and the
port will appear to be valid. If the non-daemon process just
immediately closes the connection, the Gradle client will fail
and retry 100 times.

Unfortunately, we'll keep retrying the exact same daemon because
it otherwise appears good (idle + port is connectable).

This change makes it so if we receive no response from the daemon,
we mark the daemon as unavailable and retry again. This will usually
mean that a new daemon is started or we try the next available daemon.

Fixes the underlying reason for https://github.com/gradle/gradle/issues/17345

### CODE REVIEW

**Comment by big-guy** [comment]
@bot-gradle test this

**Comment by big-guy** [comment]
@bot-gradle test this

**Comment by big-guy** [comment]
@bot-gradle test this

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonClient.java [hunk 1]
```diff
@@ -167,13 +167,16 @@ protected BuildActionResult executeBuild(Build build, DaemonClientConnection con
             result = connection.receive();
```
Comment (donat) [inline subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonClient.java:167]: ❓ Receive will return null only if the underlying object input stream receives EOF. Shouldn't we also capture some or all exception types here?

**Review by donat — APPROVED** [review]
Looks good to me. I have only one question regarding the implementation.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonClient.java (hunk 1 above)
Comment (big-guy) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonClient.java:167]: The other types of exceptions are either immediate failures (that cause the client to build to start) or are already caught in a way to cause us to retry, I believe.

**Comment by big-guy** [comment]
@bot-gradle test this

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonClient.java (hunk 1 above)
Comment (donat) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonClient.java:167]: 👍

**Comment by big-guy** [comment]
@bot-gradle test this

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Pull request merged**